network/discovery.go

```diff
@@ -218,52 +247,114 @@ func (d *discovery) getStream(peerID peer.ID) (interface{}, error) {
 	return p.stream, nil
 }
 
-func (d *discovery) findPeersCall(peerID peer.ID) ([]*peer.AddrInfo, error) {
+func (d *discovery) findPeersCall(peerID peer.ID) ([]string, error) {
 	stream, err := d.getStream(peerID)
 	if err != nil {
 		return nil, err
 	}
 
 	clt := proto.NewDiscoveryClient(stream.(*rawGrpc.ClientConn))
 
-	resp, err := clt.FindPeers(context.Background(), &proto.FindPeersReq{Count: 16})
+	resp, err := clt.FindPeers(context.Background(), &proto.FindPeersReq{Count: defaultPeerReqCount})
 	if err != nil {
 		return nil, err
 	}
 
-	addrInfo := make([]*peer.AddrInfo, len(resp.Nodes))
-
-	for indx, node := range resp.Nodes {
-		info, err := StringToAddrInfo(node)
-		if err != nil {
-			return nil, err
-		}
-
-		addrInfo[indx] = info
-	}
-
-	return addrInfo, nil
+	return resp.Nodes, nil
 }
 
 func (d *discovery) run() {
+	peerDiscoveryTicker := time.NewTicker(peerDiscoveryInterval)
+	bootnodeDiscoveryTicker := time.NewTicker(bootnodeDiscoveryInterval)
+
+	defer func() {
+		peerDiscoveryTicker.Stop()
+		bootnodeDiscoveryTicker.Stop()
+	}()
+
 	for {
 		select {
-		case <-time.After(5 * time.Second):
-		case <-d.notifyCh:
 		case <-d.closeCh:
 			return
+		case <-peerDiscoveryTicker.C:
+			go d.handleDiscovery()
+		case <-bootnodeDiscoveryTicker.C:
+			go d.bootnodeDiscovery()
 		}
-		d.handleDiscovery()
 	}
 }
 
 func (d *discovery) handleDiscovery() {
 	// take a random peer and find peers
-	if target := d.peers.getRandomPeer(); target != nil {
-		if err := d.attemptToFindPeers(target.id); err != nil {
-			d.srv.logger.Error("failed to dial peer", "peer", target.id, "err", err)
+	if d.srv.availableOutboundConns() > 0 {
+		if target := d.peers.getRandomPeer(); target != nil {
+			if err := d.attemptToFindPeers(target.id); err != nil {
+				d.srv.logger.Error("Failed to find new peers", "peer", target.id, "err", err)
+			}
+		}
+	}
+}
+
+// bootnodeDiscovery queries a random bootnode for new peers and adds them to the routing table
+func (d *discovery) bootnodeDiscovery() {
+	if d.srv.availableOutboundConns() <= 0 {
+		return
+	}
+
+	// get a random bootnode which is not connected
+	bootnode := d.srv.getBootNode()
+	if bootnode == nil {
+		return
+	}
+
+	// isTemporaryDial maintains the dial status
+	var isTemporaryDial bool
+
+	// if atleast one bootnode is connected the dial status is temporary
+	if d.bootnodeConnCount > 1 {
+		isTemporaryDial = true
+	}
```

Comment:
why don't you assign value to isTemporaryDial directly?
```suggestion
         // if atleast one bootnode is connected the dial status is temporary
	isTemporaryDial := d.bootnodeConnCount > 1
```

Reply:
I felt like, it would be more readable &  idiomatic to separate declaration with condition.